### PR TITLE
[TG-3581] Better Support for Enums in JBMC

### PR DESCRIPTION
**Pull request by mgudemann**
Introduce secific unwind bounds for copy loops in array `clone` for enumerations.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 8ca52901f74c86296a94cf7c041455f51c3e1a36).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78520669
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: c5690b25aa44e71da26a4ed08e4a01390b4fea86).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78520865
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 1fe1e3790b249a197a57078de627ca68f3d0c9d9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78577776
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 1]
```diff
@@ -2197,7 +2212,7 @@ void java_bytecode_convert_methodt::convert_invoke(
   //   generate code that may wrongly assume that such a method is
   //   accessible if we assume that its access attribute is "more
   //   accessible" than it actually is.
-  irep_idt id = arg0.get(ID_identifier);
+  const irep_idt id = arg0.get(ID_identifier);
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: `const &`?

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 2]
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
...
+  symbol.base_name = symbol_type.get(ID_C_base_name);
+  symbol.is_type = true;
+  symbol.type = class_type;
+  symbol.mode = ID_java;
+  symbol_table.add(symbol);
+
+  // Also provide a clone method:
+  // ----------------------------
+
+  const irep_idt clone_name = identifier + +".clone:()Ljava/lang/Object;";
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_utils.cpp]: `+ +`?

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 3]
```diff
@@ -5,7 +5,6 @@
 Author: Daniel Kroening, kroening@kroening.com
 
 \*******************************************************************/
-
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_utils.cpp]: spurious change?

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 4]
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
   "getCurrentThreadID"
 };
 
+/// Register in the \p symbol_table a new symbol type
+/// java::array[X] where X is byte, float, int, char...
+void add_array_type(
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_utils.cpp]: maybe add doxygen for parameters

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 5]
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
...
+  const char l,
+  symbol_tablet &symbol_table,
+  const optionalt<std::string> enum_name)
+{
+  symbol_typet symbol_type = to_symbol_type(java_array_type(l).subtype());
+
+  const std::string identifier = enum_name
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_utils.cpp]: you could assign that directly to an `irep_idt`

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 1 above)
Comment (romainbrenguier) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: it's not clear for me what `arg0` represents, if you have an intuition about that can you rename the parameter?

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 6]
```diff
@@ -2106,6 +2106,21 @@ void java_bytecode_convert_methodt::convert_invoke(
...
+  const size_t class_separator = method_name.rfind('.');
+  const std::string calling_class_name = method_name.substr(0, class_separator);
+
+  std::string class_type_name = calling_class_name.substr(6);
+  std::replace(class_type_name.begin(), class_type_name.end(), '.', '/');
+  const auto &enum_clone_symbol = symbol_table.lookup_ref(calling_class_name);
+  const bool is_enum_values_clone_call =
+    enum_clone_symbol.type.get_bool(ID_enumeration) &&
+    (full_method_name ==
+     (calling_class_name + ".values:()[L" + class_type_name + ";"));
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: define is_enum_values_clone_call closer to where it is used

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 7]
```diff
@@ -2220,7 +2235,13 @@ void java_bytecode_convert_methodt::convert_invoke(
     symbol_table.add(symbol);
   }
 
-  if(is_virtual)
+  if(is_enum_values_clone_call)
+  {
+    const std::string clone_name =
+      "java::array[" + calling_class_name + "].clone:()Ljava/lang/Object;";
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: I don't think you can add to a string literal. You should use `std::string("java::array[") + ...` instead

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 8]
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
...
+  local_symbol.pretty_name = local_symbol.base_name;
+  local_symbol.type = java_reference_type(symbol_type);
+  local_symbol.mode = ID_java;
+  symbol_table.add(local_symbol);
+  const auto &local_symexpr = local_symbol.symbol_expr();
+
+  code_blockt clone_body;
+
+  code_declt declare_cloned(local_symexpr);
+  clone_body.move_to_operands(declare_cloned);
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_utils.cpp]: code_blockt has a move method

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 9]
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
...
+    plus_exprt(new_data, index_symexpr), new_data.type().subtype());
+  code_assignt copy_cell(new_cell, old_cell);
+  copy_loop.body() = copy_cell;
+
+  // End for-loop
+  clone_body.move_to_operands(copy_loop);
+
+  member_exprt new_base_class(
+    new_array, base_class_component.get_name(), base_class_component.type());
+  address_of_exprt retval(new_base_class);
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_utils.cpp]: const?

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 10]
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
...
+  inc.operands().resize(2);
+  inc.op0() = index_symexpr;
+  inc.op1() = plus_exprt(index_symexpr, from_integer(1, index_symexpr.type()));
+  copy_loop.iter() = inc;
+
+  dereference_exprt old_cell(
+    plus_exprt(old_data, index_symexpr), old_data.type().subtype());
+  dereference_exprt new_cell(
+    plus_exprt(new_data, index_symexpr), new_data.type().subtype());
+  code_assignt copy_cell(new_cell, old_cell);
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_utils.cpp]: const?

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 11]
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
...
+  clone_symbol.pretty_name = identifier + ".clone:()";
+  clone_symbol.base_name = "clone";
+  clone_symbol.type = clone_type;
+  clone_symbol.value = clone_body;
+  clone_symbol.mode = ID_java;
+  symbol_table.add(clone_symbol);
+}
+
+/// Register in the \p symbol_table a new symbol for an enum type
+/// \param enum_type_name: Name of the enum type.
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_utils.cpp]: say that it contains the "java::" prefix?

File: jbmc/src/java_bytecode/java_enum_array_clone_unwind_handler.cpp [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+/// \file
+/// Unwind loops in array.clone for enums
+#include "java_enum_array_clone_unwind_handler.h"
+
+#include <util/invariant.h>
+#include <util/suffix.h>
+
+#include <java_bytecode/java_utils.h>
+
+/// Unwind handler that special-cases the clone functions of arrays of
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_enum_array_clone_unwind_handler.cpp]: specializes?

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 13]
```diff
@@ -2217,7 +2235,13 @@ void java_bytecode_convert_methodt::convert_invoke(
     symbol_table.add(symbol);
   }
 
-  if(is_virtual)
+  if(is_enum_values_clone_call)
+  {
+    const std::string clone_name =
+      "java::array[" + class_type_name + "].clone:()Ljava/lang/Object;";
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: I don't think you can add to a string literal

**Review by smowton — CHANGES_REQUESTED** [review]
It strikes me this can be done more simply if we can expose a bit of symex's state to the loop unwind handlers (called here https://github.com/diffblue/cbmc/blob/develop/src/cbmc/symex_bmc.cpp#L119, and implemented here https://github.com/diffblue/cbmc/blob/develop/jbmc/src/jbmc/jbmc_parse_options.cpp#L508)

Basically we want to inspect a little bit of the calling context, to check whether we're in `clone()`, and if so whether `this->length` (the length of the array being cloned) is a constant.

Possible approaches:
* Give the unwind handler the ability to call `symext::clean_expr` (which among other things performs constant propagation), so it can directly ask symex for `this->length` and find out whether we know a bound
* Annotate the loop! When generating `.clone()` we know the actual loop length is bounded and that bound is `this->length`. Then symex can evaluate the given bound and give it to the unwind handler.

I think either of these would be much simpler and cleaner than method cloning just so we have a unique function ID key to associate with the decision to unwind.

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 8 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/java_utils.cpp]: will change, but this is old code that just moved

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 13 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: I am not sure I understand what you mean, it seems to work without a problem.

**Comment by mgudemann** [comment]
@smowton We decided to go with this solution, look at the outcome and look at the alternative in a new issue.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: d449275c7a8077f606204bccdaf30156d5e10475).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78928253
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by tautschnig — COMMENTED** [review]
I acknowledge that this is work-in-progress, but just to jot down my expectations anyway: 1) there need to be tests added; 2) a change to the interpreter has been added, but it's not clear to me how that relates to the PR subject.

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 14]
```diff
@@ -445,3 +445,224 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
...
+  index_symbol.name = index_name;
+  index_symbol.base_name = "index";
+  index_symbol.pretty_name = index_symbol.base_name;
+  index_symbol.type = length_component.type();
+  index_symbol.mode = ID_java;
+  symbol_table.add(index_symbol);
+  const auto &index_symexpr = index_symbol.symbol_expr();
+
+  code_declt declare_index(index_symexpr);
+  clone_body.move_to_operands(declare_index);
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_utils.cpp]: clone_body.add(code_declt(...

File: jbmc/src/java_bytecode/java_utils.cpp [hunk 15]
```diff
@@ -445,3 +445,224 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
...
+
+  side_effect_exprt inc(ID_assign);
+  inc.operands().resize(2);
+  inc.op0() = index_symexpr;
+  inc.op1() = plus_exprt(index_symexpr, from_integer(1, index_symexpr.type()));
+  copy_loop.iter() = inc;
+
+  dereference_exprt old_cell(
+    plus_exprt(old_data, index_symexpr), old_data.type().subtype());
+  dereference_exprt new_cell(
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_utils.cpp]: const?

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 14 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/java_utils.cpp]: done

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 15 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/java_utils.cpp]: done

please note that this code was just moved, not written by me

Comment (mgudemann) [reply jbmc/src/java_bytecode/java_utils.cpp]: done, please note that I mainly moved the code and didn't write it

File: jbmc/src/java_bytecode/java_utils.cpp (hunk 14 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/java_utils.cpp]: done

**Comment by mgudemann** [comment]
@tautschnig please find some regression tests in the latest commit (these were just in TG before)
The interpreter change is to prevent a warning / error message that occurred

**Comment by tautschnig** [comment]
> The interpreter change is to prevent a warning / error message that occurred

Would it somehow be possible to explain what this is? The change seems entirely sensible to me, but I still don't really get why it would fit this PR.

**Comment by mgudemann** [comment]
@tautschnig this fixes the interpreter when it got a `member-of-constant-struct` like 
`void*[{.length = 6, .data = &dynamic_7_array}.length]`

**Comment by mgudemann** [comment]
these occurred with my PR, but I am not sure whether it was the cause or I just paid more attention to the output here

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 9b8237777a037315cb84773756f22efde7024169).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80539334
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Comment by smowton** [comment]
Proposed alternative implementation here: https://github.com/diffblue/cbmc/pull/2651

**Comment by mgudemann** [comment]
superseded by #2651

**Pull request closed without merging**